### ISSUE TITLE
Generic class with a single upper bound of a union type 

### ISSUE THREAD

**Issue by ritsuki1227** [python/mypy#18313]
**Bug Report**

When implementing the following generic class with a single upper bound of a union type, you get the following errors, while no errors are raised when using multiple upper-bounds.

**To Reproduce**

```python

class Example1[T: (str, int)]:
    def foo(self, a: T) -> T:
        if isinstance(a, int):
            return a + 1
        elif isinstance(a, str):
            return a + "hello"


class Example2[T: int | str]:
    def bar(self, a: T) -> T:
        if isinstance(a, int):
            return a + 1
        elif isinstance(a, str):
            return a + "hello"
```

**Expected Behavior**

Class `Example2` should raise no errors.

**Actual Behavior**

Class `Example2` raises the following error: 
```
error: Incompatible return value type (got "int", expected "T")  [return-value]
error: Incompatible return value type (got "str", expected "T")  [return-value]
```

**Your Environment**

- Mypy version used: 1.13.0
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12.3

**Comment by JelleZijlstra** [comment]
This error is correct.

Your Example1 does not use multiple bounds, but constraints, which work differently. A TypeVar with constraints can be solved only to exactly `int` or `str`, and the method must return `int` if an `int` is given, and `str` if a `str` is given. But Example2 uses a bound, and the signature says that the return type must be exactly the same as the input type, for any consistent subtype of `int | str`. For example, `bool` is a subclass of `int`, so `Example2().bar(True)` should return `bool`, but in your implementation it doesn't: it returns 2, an int.

**Comment by ritsuki1227** [comment]
@JelleZijlstra Thank you for your clear explanation even though it is an expected behavior. Your examples are also easy to understand.
